django/contrib/auth/__init__.py

```diff
@@ -37,6 +36,10 @@ def get_backends():
     return _get_backends(return_tuples=False)
 
 
+def csrf_is_enabled():
+    return 'django.middleware.csrf.CsrfViewMiddleware' in settings.MIDDLEWARE
```

Comment:
This will perform a full scan over a list and doesn't handle `CsrfViewMiddleware` subclassing.

Reply:
Done

Reply:
Done

> I'm not sure I understand the necessity of having `login` only call `rotate_token` if the middleware is installed.

In this case it would be for not annotate the request in-necessarily. Well, it can be executed and left there anyway, it may be not ideal (to have inactive middleware "leftovers"), but for performance and simplicity (assuming the check is costly and is on a sensitive function) could be done.

IMO, ideally, login() should only "Persist a user id and a backend in the request" and set a new_login flag or so for later. The middleware should catch it and do its logic, like rotate the token or flush the session. This way, there would be no middleware logic in the function and no need to check, but it would require non-trivial refactoring (if not major) and it would be out of the scope or goal of this PR (this would be the first step into that direction for the CSRF MW) - though can be reviewed if you agree.

> 
> Can't we simply branch of `'CSRF_COOKIE' in request.META` in `rotate_token` given this value is only set if `CsrfViewMiddleware` is installed?
> 
> ```python
> def rotate_token(request):
>     """
>     Marks a request for CSRF token rotation by the middlware - should be done
>     on login for security purposes.
>     """
>     # No need to rotate token if it doesn't exist yet.
>     if 'CSRF_COOKIE' not in request.META:
>         return

If there no CSRF_COOKIE in request.META, the middleware will reject the request and not process the view (CSFR_MW.process_view()). If it gets up to here, there will be a valid CSRF token or CSRF has been disabled. Ahhh... do you mean using this as check for the csfr middleware enablement? It could, but if someone disables it only for the login view, then, it would not rotate on login, it would still use the one generated previously.

>     request.META['CSRF_COOKIE_USED'] = True
>     request.csrf_cookie_needs_reset = True
>     request.csrf_token_rotation = True
> ```